python-backend/math_utils.py

```diff
@@ -12,91 +14,150 @@
-    :param symbol: the variable used by the user in numerator and denominator
 
     Returns
     -------
     list of values computed at each step from n = 1 on to some n max
     """
-    a_n = [mpmath.mpf(a.subs(symbol, n).evalf(PRECISION)) for n in range(0, iterations)]
+    a_n = [a(n) for n in range(0, 2 * iterations)]
     for i in range(0, min(constants.DEBUG_LINES, len(a_n))):
         logger.debug(f"a value at {i}: {a_n[i]}")
 
-    # these arrays start at n = -2, so we prepad the convergent list with values we will slice off
-    numerators = [mpmath.mpf(0), mpmath.mpf(1)]
-    denominators = [mpmath.mpf(1), mpmath.mpf(0)]
-    convergents = [None, None]
+    # these arrays start at n = -2
+    numerators = OrderedDict({-2: mpmath.mpf(0), -1: mpmath.mpf(1)})
+    denominators = OrderedDict({-2: mpmath.mpf(1), -1: mpmath.mpf(0)})
+    convergents = OrderedDict({-2: None, -1: None})
 
-    for i in range(0, len(a_n)):
-        num = a_n[i] * numerators[i - 1 + 2] + numerators[i - 2 + 2]
-        numerators.append(num)
-        denom = a_n[i] * denominators[i - 1 + 2] + denominators[i - 2 + 2]
-        denominators.append(denom)
+    for n in range(0, 2 * iterations):
+        numerators[n] = a_n[n] * numerators[n - 1] + numerators[n - 2]
+        denominators[n] = a_n[n] * denominators[n - 1] + denominators[n - 2]
 
-        if denominators[i] != 0:
-            convergents.append(numerators[-1] / denominators[-1])
-            if i < constants.DEBUG_LINES:
+        try:
+            convergents[n] = (numerators[n] / denominators[n])
+            if n < constants.DEBUG_LINES:
                 logger.debug(
-                    f"simple computed values n: {i} num: {numerators[-1]} denom: {denominators[-1]} ratio: {convergents[-1]}")
-        else:
-            convergents.append(None)
+                    f"simple computed values n: {n} num: {numerators[n]} denom: {denominators[n]} "
+                    f"ratio: {convergents[n]}")
+        except ZeroDivisionError:
             logger.warning(
-                f"simple computed values n: {i} num: {numerators[-1]} denom: {denominators[-1]} ratio: Undefined")
+                f"simple computed values n: {n} num: {numerators[n]} denom: {denominators[n]} ratio: Undefined")
+            break
+
+    return convergents, numerators, denominators
 
-    # return from n = 0 on
-    return convergents[2:], denominators[2:]
+
+def laurent(a: Number, b: Number, symbol: Symbol = None, term_count=2) -> str:
+    """
+    compute laurent series in symbol up to term_count terms
+    """
+    result = 0
+    power = 0
+    logger.debug(f"terms: {term_count}")
+
+    expression = 1 + 4 * b / (a * a.subs(symbol, symbol - 1)) if symbol is not None else 1 + 4 * b / (a * a)
+    logger.debug(f"initial expression: {expression}")
+    logger.debug(f"limit {limit_seq(expression)}")
+
+    while limit_seq(expression if symbol is None else expression / symbol ** power).is_infinite:
```

Comment:
It doesn't look like `symbol` changes in the body of this function, so it might be helpful to calculate functions that run the `symbol`-related calculations outside of the loops.

For example...

```
if symbol is None:
    power_calc_func = lambda: expression
    term_count_func = lambda: limit_seq(expression)
else:
    power_calc_func = lambda: expression / symbol ** power
    term_count_func = lambda: limit_seq(expression / symbol ** power)
```

After writing this, it looks like the lambda funcs could be even simpler than what I have written here.